File: app/src/main/java/com/example/e_commercekotlin/presentation/screens/FollowFragment.java
package com.example.e_commercekotlin.presentation.screens;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;

import com.example.e_commercekotlin.R;
import com.example.e_commercekotlin.databinding.FragmentFollowBinding;
import com.example.e_commercekotlin.presentation.adapter.FollowingAdapter;
import com.example.e_commercekotlin.presentation.model.Featured;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;
import java.util.List;

public class FollowFragment extends Fragment {

    private FragmentFollowBinding bindingFollow;
    private FollowingAdapter followingAdapter;
    private List<Featured> features = new ArrayList<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        bindingFollow = FragmentFollowBinding.inflate(inflater, container, false);
        View view = bindingFollow.getRoot();
        bindingFollow.followButton.buttonTv.setText("Follow");
        bindingFollow.followButton.buttonTv.setTextColor(ContextCompat.getColor(bindingFollow.followButton.getRoot().getContext(), R.color.white));


        bindingFollow.followButton.getRoot().setOnClickListener(new View.OnClickListener() {
            private boolean isFollowing = false;

            @Override
            public void onClick(View v) {
                if (isFollowing) {
                    bindingFollow.followButton.buttonTv.setText("Follow");
                    bindingFollow.followButton.getRoot().setBackgroundResource(R.drawable.follow);
                    bindingFollow.followButton.buttonTv.setTextColor(ContextCompat.getColor(bindingFollow.followButton.getRoot().getContext(), R.color.white));
                    Snackbar.make(view, "Sustainable removed from your feed tabs", Snackbar.LENGTH_SHORT).show();
                } else {
                    bindingFollow.followButton.buttonTv.setText("Following");
                    bindingFollow.followButton.buttonTv.setTextColor(ContextCompat.getColor(bindingFollow.followButton.getRoot().getContext(), R.color.black));
                    bindingFollow.followButton.getRoot().setBackgroundResource(R.drawable.following);  // Change to the grey drawable background
                    Snackbar.make(view, "Sustainable added to your feed tabs", Snackbar.LENGTH_SHORT).show();
                }
                isFollowing = !isFollowing;
            }
        });

        TabLayout tabLayout = view.findViewById(R.id.tabLayout);
        tabLayout.addTab(tabLayout.newTab().setText("Products"));
        tabLayout.addTab(tabLayout.newTab().setText("Collections"));
        tabLayout.addTab(tabLayout.newTab().setText("Stores"));

        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        followingAdapter = new FollowingAdapter();
        bindingFollow.followrecycle.setLayoutManager(new GridLayoutManager(getContext(), 2));
        bindingFollow.followrecycle.setAdapter(followingAdapter);

        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));
        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));
        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));
        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));
        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));
        features.add(new Featured(R.drawable.baseline_profile_24, "T-shirt", "too tight", "200"));


        followingAdapter.setFeatures(features);
    }
}

File: app/src/main/java/com/example/e_commercekotlin/presentation/screens/FollowFragment.kt
package com.example.e_commercekotlin.presentation.screens

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.example.e_commercekotlin.Util.Constants
import com.example.e_commercekotlin.Util.hide
import com.example.e_commercekotlin.Util.setBottomNavVisibility
import com.example.e_commercekotlin.Util.show
import com.example.e_commercekotlin.data.Resource
import com.example.e_commercekotlin.data.model.toProductItem
import com.example.e_commercekotlin.databinding.FragmentFollowBinding
import com.example.e_commercekotlin.presentation.adapter.ProductAdapter
import com.example.e_commercekotlin.presentation.viewmodels.ProductViewModel

class FollowFragment : Fragment(), ProductAdapter.ClickListener {

    private lateinit var binding: FragmentFollowBinding
    private val args: FollowFragmentArgs by navArgs()
    private val productsViewModel: ProductViewModel by viewModels()
    private lateinit var productAdapter: ProductAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentFollowBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        activity?.setBottomNavVisibility(visible = false)

        val imageResId = args.imageResId
        val title = args.title

        productAdapter = ProductAdapter()
        binding.followrecycle.layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        binding.followrecycle.adapter = productAdapter
        observeProducts()

        binding.followImage.setImageResource(imageResId)
        binding.followTitle.text = title


        binding.followButton.buttonTv.text = Constants.FOLLOW
        binding.followButton.buttonTv.setOnClickListener {
            toggleFollowState()
        }

        productAdapter.setListener(this)
//
//        productAdapter.setListener(object : ProductAdapter.ClickListener {
//            override fun onProductClick(productId: Long, productName: String, productImage: String) {
//
//            }
//        })


    }

    private fun toggleFollowState() {
        val currentText = binding.followButton.buttonTv.text.toString()
        binding.followButton.buttonTv.text = if (currentText == Constants.FOLLOW) "Following" else Constants.FOLLOW
    }

    private fun observeProducts() {
        productsViewModel.allProduct.observe(viewLifecycleOwner) { resource ->
            when (resource) {
                is Resource.Loading -> {
                    binding.progressBar.show()
                }
                is Resource.Success -> {
                    binding.progressBar.hide()
                    handleLoadingState(false)
                    resource.data?.let {
                        productAdapter.setProductList(it.map { it.toProductItem() })}
                }
                is Resource.Error -> {
                    binding.progressBar.hide()

                }
            }
        }
    }

    private fun handleLoadingState(isLoading: Boolean) {
        binding.progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
    }

    override fun onProductClick(productId: Long, productName: String, productImage: String) {
        // Create a Bundle with the product details
        val bundle = Bundle().apply {
            putInt("productId", productId.toInt())
            putString("product_name", productName)
            putString("product_image", productImage)
            putString("source_fragment", "FeedFragment")

        }

        // Create an instance of CustomDialogFragment and pass the navigation action as a lambda
        val dialogFragment = CustomDialogFragment {
            // Action to be executed when the user clicks in the dialog
            val action = FollowFragmentDirections.actionFollowFragmentToProductDetails(productId.toInt())
            findNavController().navigate(action)
        }

        // Set the arguments (product details) for the dialog
        dialogFragment.arguments = bundle

        // Show the dialog
        dialogFragment.show(parentFragmentManager, "CustomDialogFragment")
    }
}
